### PR TITLE
Added apiServerAccessProfile object to azure_rm_aks

### LINKED ISSUE #335: [Feature Idea] azure_rm_aks - Add apiServerAccessProfile option
##### SUMMARY
Currently, the azure_rm_aks module does not feature the ability to specify the apiServerAccessProfile for an aks cluster.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
azure_rm_aks

##### ADDITIONAL INFORMATION

https://docs.microsoft.com/en-us/azure/templates/microsoft.containerservice/managedclusters#managedclusterproperties-object

- This would allow users to create a private cluster through this module.
- This would also allow users to specify an authorized IP range to restrict API access to.

### PR DESCRIPTION
**Pull request by MarcGi**
##### SUMMARY
Fixes: #335 , #298 , #435

Added option to specify the apiServerAccessProfile through the existing azure_rm_aks module

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

azure_rm_aks

##### ADDITIONAL INFORMATION

```paste below
- name: Create an AKS instance
      azure_rm_aks:
        name: myAKS
        resource_group: myResourceGroup
        location: eastus
        dns_prefix: akstest
        kubernetes_version: 1.18.0
        linux_profile:
          admin_username: azureuser
          ssh_key: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAA...
        service_principal:
          client_id: "cf72ca99-f6b9-4004-b0e0-bee10c521948"
          client_secret: "Password1234!"
        api_server_access_profile:
          authorized_ip_ranges: 
            - "192.0.2.0"
            - "198.51.100.0"
            - "203.0.113.0"
          enable_private_cluster: yes
        agent_pool_profiles:
          - name: default
            count: 1
            vm_size: Standard_DS1_v2
            enable_auto_scaling: True
            type: VirtualMachineScaleSets
            max_count: 3
            min_count: 1
            enable_rbac: yes
```

### CODE REVIEW

**Comment by MarcGi** [comment]
Hi @Fred-sun is it possible that you could review this pr when you get the time, cheers!

**Comment by Fred-sun** [comment]
@MarcGi 
```e.g
ERROR: plugins/modules/azure_rm_aks.py:201:29: W291: trailing whitespace
ERROR: plugins/modules/azure_rm_aks.py:403:1: E302: expected 2 blank lines, found 1
ERROR: plugins/modules/azure_rm_aks.py:406:1: E302: expected 2 blank lines, found 1
ERROR: plugins/modules/azure_rm_aks.py:540:1: E302: expected 2 blank lines, found 1
ERROR: plugins/modules/azure_rm_aks.py:710:161: E501: line too long (178 > 160 characters)
ERROR: plugins/modules/azure_rm_aks.py:713:161: E501: line too long (178 > 160 characters)
ERROR: plugins/modules/azure_rm_aks.py:715:1: W293: blank line contains whitespace
ERROR: plugins/modules/azure_rm_aks.py:958:5: E112: expected an indented block

ERROR: plugins/modules/azure_rm_aks.py:382:13: syntax-error: invalid syntax (<unknown>, line 382)

ERROR: plugins/modules/azure_rm_aks.py:0:0: python-syntax-error: Python SyntaxError while parsing module
ERROR: tests/sanity/ignore-2.10.txt:1:1: A100: Ignoring 'doc-choices-do-not-match-spec' on 'plugins/modules/azure_rm_aks.py' is unnecessary
ERROR: tests/sanity/ignore-2.10.txt:2:1: A100: Ignoring 'doc-default-does-not-match-spec' on 'plugins/modules/azure_rm_aks.py' is unnecessary
ERROR: tests/sanity/ignore-2.10.txt:3:1: A100: Ignoring 'parameter-type-not-in-doc' on 'plugins/modules/azure_rm_aks.py' is unnecessary
ERROR: tests/sanity/ignore-2.10.txt:4:1: A100: Ignoring 'undocumented-parameter' on 'plugins/modules/azure_rm_aks.py' is unnecessary
ERROR: tests/sanity/ignore-2.10.txt:5:1: A100: Ignoring 'doc-elements-mismatch' on 'plugins/modules/azure_rm_aks.py' is unnecessary
ERROR: tests/sanity/ignore-2.10.txt:6:1: A100: Ignoring 'invalid-ansiblemodule-schema' on 'plugins/modules/azure_rm_aks.py' is unnecessary
ERROR: tests/sanity/ignore-2.10.txt:7:1: A100: Ignoring 'required_if-requirements-unknown' on 'plugins/modules/azure_rm_aks.py' is unnecessary
ERROR: tests/sanity/ignore-2.10.txt:8:1: A100: Ignoring 'required_if-unknown-key' on 'plugins/modules/azure_rm_aks.py' is unnecessary

**Comment by MarcGi** [comment]
Will address sanity errors in another commit

**Comment by MarcGi** [comment]
Hi @Fred-sun, can you confirm how you are running sanity tests?

As when I am running them against the aks module I get the following:

```bash 
ansible-test sanity --test validate-modules --docker default plugins/modules/azure_rm_aks.py
```

```python
ERROR: Found 67 validate-modules issue(s) which need to be resolved:
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-choices-do-not-match-spec: Argument 'load_balancer_sku' in argument_spec found in network_profile defines choices as ([]) but documentation defines choices as (['standard', 'basic'])
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-choices-do-not-match-spec: Argument 'network_policy' in argument_spec found in network_profile defines choices as ([]) but documentation defines choices as (['azure', 'calico'])
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-choices-do-not-match-spec: Argument 'os_type' in argument_spec found in agent_pool_profiles defines choices as (['Linux', 'Windows']) but documentation defines choices as ([])
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-choices-do-not-match-spec: Argument 'storage_profiles' in argument_spec found in agent_pool_profiles defines choices as (['StorageAccount', 'ManagedDisks']) but documentation defines choices as ([])
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-choices-do-not-match-spec: Argument 'type' in argument_spec found in agent_pool_profiles defines choices as ([]) but documentation defines choices as (['VirtualMachineScaleSets', 'AvailabilitySet'])
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-default-does-not-match-spec: Argument 'dns_service_ip' in argument_spec found in network_profile defines default as (None) but documentation defines default as ('10.0.0.10')
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-default-does-not-match-spec: Argument 'docker_bridge_cidr' in argument_spec found in network_profile defines default as (None) but documentation defines default as ('172.17.0.1/16')
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-default-does-not-match-spec: Argument 'enabled' in argument_spec found in addon -> http_application_routing defines default as (True) but documentation defines default as (False)
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-default-does-not-match-spec: Argument 'enabled' in argument_spec found in addon -> monitoring defines default as (True) but documentation defines default as (False)
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-default-does-not-match-spec: Argument 'enabled' in argument_spec found in addon -> virtual_node defines default as (True) but documentation defines default as (False)
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-default-does-not-match-spec: Argument 'network_plugin' in argument_spec found in network_profile defines default as (None) but documentation defines default as ('kubenet')
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-default-does-not-match-spec: Argument 'pod_cidr' in argument_spec found in network_profile defines default as (None) but documentation defines default as ('10.244.0.0/16')
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-default-does-not-match-spec: Argument 'service_cidr' in argument_spec found in network_profile defines default as (None) but documentation defines default as ('10.0.0.0/16')
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-elements-mismatch: Argument 'agent_pool_profiles' in argument_spec specifies elements as dict,but elements is not documented
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-elements-mismatch: Argument 'authorized_ip_ranges' in argument_spec found in api_server_access_profile specifies elements as string,but elements is not documented
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-elements-mismatch: Argument 'availability_zones' in argument_spec found in agent_pool_profiles specifies elements as int,but elements is not documented
ERROR: plugins/modules/azure_rm_aks.py:0:0: doc-elements-mismatch: Argument 'ports' in argument_spec found in agent_pool_profiles specifies elements as int,but elements is not documented
ERROR: plugins/modules/azure_rm_aks.py:0:0: invalid-ansiblemodule-schema: AnsibleModule.argument_spec.agent_pool_profiles.options.type.choice: extra keys not allowed @ data['argument_spec']['agent_pool_profiles']['options']['type']['choice']. Got ['VirtualMachineScaleSets', 'AvailabilitySet']
ERROR: plugins/modules/azure_rm_aks.py:0:0: invalid-ansiblemodule-schema: AnsibleModule.argument_spec.api_server_access_profile.options.authorized_ip_ranges.elements: not a valid value for dictionary value @ data['argument_spec']['api_server_access_profile']['options']['authorized_ip_ranges']['elements']. Got 'string'
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'aad_profile' in argument_spec defines type as 'dict' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'admin_username' in argument_spec found in linux_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'agent_pool_profiles' in argument_spec defines type as 'list' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'api_server_access_profile' in argument_spec defines type as 'dict' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'client_app_id' in argument_spec found in aad_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'client_id' in argument_spec found in service_principal defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'client_secret' in argument_spec found in service_principal defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'count' in argument_spec found in agent_pool_profiles defines type as 'int' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'dns_prefix' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'dns_prefix' in argument_spec found in agent_pool_profiles defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'dns_service_ip' in argument_spec found in network_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'docker_bridge_cidr' in argument_spec found in network_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'kubernetes_version' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'linux_profile' in argument_spec defines type as 'dict' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'load_balancer_sku' in argument_spec found in network_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'location' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'log_analytics_workspace_resource_id' in argument_spec found in addon -> monitoring defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'name' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'name' in argument_spec found in agent_pool_profiles defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'network_plugin' in argument_spec found in network_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'network_policy' in argument_spec found in network_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'network_profile' in argument_spec defines type as 'dict' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'os_disk_size_gb' in argument_spec found in agent_pool_profiles defines type as 'int' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'os_type' in argument_spec found in agent_pool_profiles defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'pod_cidr' in argument_spec found in network_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'ports' in argument_spec found in agent_pool_profiles defines type as 'list' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'resource_group' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'server_app_id' in argument_spec found in aad_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'server_app_secret' in argument_spec found in aad_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'service_cidr' in argument_spec found in network_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'service_principal' in argument_spec defines type as 'dict' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'ssh_key' in argument_spec found in linux_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'state' in argument_spec defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'storage_profiles' in argument_spec found in agent_pool_profiles defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'subnet_resource_id' in argument_spec found in addon -> virtual_node defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'tenant_id' in argument_spec found in aad_profile defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: parameter-type-not-in-doc: Argument 'vm_size' in argument_spec found in agent_pool_profiles defines type as 'str' but documentation doesn't define type
ERROR: plugins/modules/azure_rm_aks.py:0:0: required_if-requirements-unknown: required_if contains terms in requirements which are not part of argument_spec: log_path
ERROR: plugins/modules/azure_rm_aks.py:0:0: required_if-unknown-key: required_if must have its key log_mode in argument_spec
ERROR: plugins/modules/azure_rm_aks.py:0:0: undocumented-parameter: Argument 'SubnetName' found in addon -> virtual_node is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_aks.py:0:0: undocumented-parameter: Argument 'aciConnector' found in addon is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_aks.py:0:0: undocumented-parameter: Argument 'dns_prefix' found in agent_pool_profiles is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_aks.py:0:0: undocumented-parameter: Argument 'httpApplicationRouting' found in addon is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_aks.py:0:0: undocumented-parameter: Argument 'logAnalyticsWorkspaceResourceID' found in addon -> monitoring is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_aks.py:0:0: undocumented-parameter: Argument 'omsagent' found in addon is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_aks.py:0:0: undocumented-parameter: Argument 'os_type' found in agent_pool_profiles is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_aks.py:0:0: undocumented-parameter: Argument 'ports' found in agent_pool_profiles is listed in the argument_spec, but not documented in the module documentation
ERROR: plugins/modules/azure_rm_aks.py:0:0: undocumented-parameter: Argument 'storage_profiles' found in agent_pool_profiles is listed in the argument_spec, but not documented in the module documentation
```

Thanks!

**Comment by MarcGi** [comment]
Hi @Fred-sun, I have made the required changes, could you please review again?

Thanks

**Comment by Fred-sun** [comment]
> Hi @Fred-sun, I have made the required changes, could you please review again?
> 
> Thanks

@MarcGi Why delete items in ignore file? Please keep. Thank you very much!

**Comment by MarcGi** [comment]
Hi @Fred-sun, Apologies, I misunderstood a previous comment. I have reverted this commit. Thank You!

**Comment by Ompragash** [comment]
@Fred-sun Could you kindly take a look on this PR?

**Comment by Fred-sun** [comment]
> @Fred-sun Could you kindly take a look on this PR?

@Ompragash I'm working in this, will push for review and merge! Thank you very much!

File: plugins/modules/azure_rm_aks.py [hunk 1]
```diff
@@ -520,6 +537,12 @@ def create_addon_profiles_spec():
 )
 
 
+api_server_access_profile_spec = dict(
+    authorized_ip_ranges=dict(type='list', elements='string'),
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_aks.py]: There is no 'string' type in python!
```suggestion
    authorized_ip_ranges=dict(type='list', elements='str'),
```

File: plugins/modules/azure_rm_aks.py [hunk 2]
```diff
@@ -973,6 +1014,9 @@ def create_network_profile_instance(self, network):
     def create_aad_profile_instance(self, aad):
         return self.managedcluster_models.ManagedClusterAADProfile(**aad) if aad else None
 
+    def create_api_server_access_profile_spec(self, server_access):
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_aks.py]: The call does not match the name of function define
```suggestion
    def create_api_server_access_profile_instance(self, server_access):
```

**Review by Fred-sun — COMMENTED** [review]
@MarcGi You can share your script "tests/integration/the targets/azure_rm_aks/tasks/main yml" the result of the execution? Thank you very much!

**Comment by Fred-sun** [comment]
@MarcGi Any feedback to "tests/integration/the targets/azure_rm_aks/tasks/main yml" test? Thank you very much!

**Comment by MarcGi** [comment]
Hi @Fred-sun, I will be able to action this around 6pm Utc today, just to check are you making use of the ansible-test integration tool or building the module and installing it?

Thank you very much!

**Comment by MarcGi** [comment]
Hi @Fred-sun, I have got the integration test running and I have discovered a few bugs. The main one is the ContainerServiceClient API version needs (in azure_rm_common.py) to be changed from '2019-04-01'  to '2019-08-01' which will introduce a test failure when running azure_rm_aks/main.yml tests: (regarding azure_rm_aksupgrade_info):

```
TASK [azure_rm_aks : Get AKS upgrade versions] ******************************************************************************************************************************************************************
ok: [testhost] => {"azure_aks_upgrades": {"agent_pool_profiles": null, "control_plane_profile": {"kubernetes_version": "1.17.13", "name": null, "os_type": "Linux", "upgrades": [{"is_preview": null, "kubernetes_version": "1.17.16"}, {"is_preview": null, "kubernetes_version": "1.18.10"}, {"is_preview": null, "kubernetes_version": "1.18.14"}]}}, "changed": false}

TASK [azure_rm_aks : Assert available versions for upgrade] *****************************************************************************************************************************************************
fatal: [testhost]: FAILED! => {"msg": "The conditional check 'upgrades.azure_aks_upgrades.agent_pool_profiles[0].kubernetes_version == versions.azure_aks_versions[0]' failed. The error was: error while evaluating conditional (upgrades.azure_aks_upgrades.agent_pool_profiles[0].kubernetes_version == versions.azure_aks_versions[0]): None has no element 0"}
```

The portal also shows the following message when trying to upgrade the AKS cluster manually:
```
This node pool is using the latest version of Kubernetes supported by the cluster. To upgrade to a later version of Kubernetes, upgrade the cluster control plane first in the 'Upgrade' menu item
```

I will start working through the required changes soon if that is good with you? :) 

Thank You!

**Comment by Fred-sun** [comment]
@MarcGi Yes ！It's can update to "2020-04-01", Thank you very much!
```e,g,
https://github.com/ansible-collections/azure/blob/9e20c6e83e7b9ff75293db8d0a28921a2fa88ff6/plugins/module_utils/azure_rm_common.py#L1129
https://github.com/ansible-collections/azure/blob/9e20c6e83e7b9ff75293db8d0a28921a2fa88ff6/plugins/module_utils/azure_rm_common.py#L1121

**Comment by MarcGi** [comment]
Hi @Fred-sun, updating the API version to 2020-04-01 results in:

"msg": "Error creating the AKS instance: Operation failed with status: 'Bad Request'. Details: Must define at least one system pool."
More details can be found here: https://github.com/Azure/AKS/issues/1602

I would suggest we create a branch and PR for updating the API version to 2020-04-1 and deal with any breaking changes before fixing this PR. Does that sound good to you? Or would you prefer to have all these changes in this branch/PR?

Thank You Very Much!

**Comment by Fred-sun** [comment]
@MarcGi Agreed, thank you for your contribution!

**Comment by Fred-sun** [comment]
@MarcGi Thank you for your contribution. Can you help resolve conflicting files? I will move forward with the merger as soon as possible. Thank you very much!

**Comment by Fred-sun** [comment]
@MarcGi  Could you please help to authorize me to update this PR? Thank you very much!

**Comment by MarcGi** [comment]
Hi @Fred-sun, hopefully I will have some time this week to resolve these files. Thank you!

**Comment by Fred-sun** [comment]
@MarcGi Any update for this? Thank you very much!

**Comment by Fred-sun** [comment]
duplicate with #651 ,

**Pull request closed without merging**